hazelcast/src/test/java/com/hazelcast/client/AuthenticationInformationLeakTest.java

```diff
@@ -0,0 +1,198 @@
+    }
+
+    @Test
+    public void testFailedAuthenticationDoesNotLeakInfoClientNotAllowed() throws Exception {
+        // No client is allowed in the cluster
+        getClientEngineImpl(instance).applySelector(ClientSelectors.none());
+        authenticateAndAssert(AuthenticationStatus.NOT_ALLOWED_IN_CLUSTER, serializationVersion, false, clusterName);
+    }
+
+    private void authenticateAndAssert(AuthenticationStatus status, byte serVersion, boolean useWrongClusterName, String clusterName) throws IOException {
+        InetSocketAddress endpoint = new InetSocketAddress("127.0.0.1", 5701);
+        try (Socket socket = new Socket()) {
+            socket.setReuseAddress(true);
+            socket.connect(endpoint);
+            try (OutputStream os = socket.getOutputStream(); InputStream is = socket.getInputStream()) {
+                os.write(CLIENT_BINARY.getBytes(StandardCharsets.UTF_8));
+                String clientClusterName = useWrongClusterName ? clusterName + 'a' : clusterName;
+                ClientMessage res = authenticate(clientClusterName, serVersion, os, is);
+                ClientAuthenticationCodec.ResponseParameters responseParameters = ClientAuthenticationCodec.decodeResponse(res);
+                assertEquals(status.getId(), responseParameters.status);
+                assertEquals(-1, responseParameters.partitionCount);
+                assertEquals(-1, responseParameters.serializationVersion);
+                assertEquals("", responseParameters.serverHazelcastVersion);
+                assertNull(responseParameters.address);
+                assertNull(responseParameters.memberUuid);
+                assertNull(responseParameters.clusterId);
+            }
+        }
+    }
+
+    private ClientMessage authenticate(String clusterName, byte serVersion, OutputStream os, InputStream is)
+            throws IOException {
+        UUID uuid = new UUID(0, 0);
+        ClientMessage msg = ClientAuthenticationCodec.encodeRequest(clusterName, null, null, uuid, "", serVersion, "", "", new ArrayList<>());
+        ClientTestUtil.writeClientMessage(os, msg);
+        return readResponse(is, ClientAuthenticationCodec.RESPONSE_MESSAGE_TYPE);
+    }
+
+    private void getKeyValue(SerializationService ss, OutputStream os, InputStream is)
+            throws IOException {
+        Data keyData = ss.toData("key");
+        ClientMessage msg = MapGetCodec.encodeRequest("mapName", keyData, 0);
+        msg.setPartitionId(getPartitionId(keyData));
+        ClientTestUtil.writeClientMessage(os, msg);
+        readResponse(is, MapGetCodec.RESPONSE_MESSAGE_TYPE);
+    }
+
+    private int getPartitionId(Data keyData) {
+        int hash = keyData.getPartitionHash();
+        return HashUtil.hashToIndex(hash, 271);
+    }
+
+    private ClientMessage readResponse(InputStream is, int expectedMsgType) throws IOException {
+        ClientMessage clientMessage = ClientMessage.createForEncode();
+        int msgType;
+        do {
```

Comment:
why loop here?

Why don't you check for response `EXCEPTION_MESSAGE_TYPE == message.getMessageType()` and then 
 use `client.getClientExceptionFactory().createException(message)` to parse the response?

Reply:
This part is from Josef's attacker code actually. I think looping is needed because there may be more than one frames.

`EXCEPTION_MESSAGE_TYPE == message.getMessageType()` is checked at line 191. And the error is thrown as `RuntimeException` instead of creating an exception. Maybe we can use `client.getClientExceptionFactory().createException(message)` as well. However, if we move this code to a Util file as Metin suggested, it may break some other test code. Let me see if there is such a test.

Reply:
@mdumandag Actually two code is not the same. readResponse in `phonehome/TestUtil.java` does not include a check for exception message type. See the screenshot. 

![image](https://user-images.githubusercontent.com/32355782/189915970-20dd57a8-0455-4806-ba35-51c08f0068bc.png)

Reply:
There is also another readResponse in ClientMessageProtectionTest and it is the same with the one in phonehome/TestUtil.java.

Reply:
Do you read more than one client message in any of these tests? I believe in others, we are reading just a single message and I believe the same happens here as well. So, it should be OK to remove the do while loop here, make this one same with others, put this into the utility class, and use it from there